zerver/views/message_fetch.py

```diff
@@ -912,7 +996,14 @@ def get_messages_backend(request: HttpRequest, user_profile: UserProfile,
     # 'messages' list.
     message_ids: List[int] = []
     user_message_flags: Dict[int, List[str]] = {}
-    if include_history:
+    if is_web_public_query:
+        # For web-public users, we treat all historical messages as read.
+        for row in rows:
+            message_id = row[0]
+            message_ids.append(message_id)
+            user_message_flags[message_id] = ["read"]
+    elif include_history:
+        assert user_profile is not None
```

Comment:
The possibility of having `user_profile` be `AnonymousUser` earlier in the codepath makes me think we should be more defensive here and replace all `assert user_profile is not None` with `assert isinstance(user_profile, UserProfile)` to prevent some future bugs from leaking `AnonymousUser` object down this path and causing unauthorized access to non-public messages.

Reply:
Can it be?  I think only the `maybe_user_profile` can be an AnonymousUser with the logic.

Reply:
I think mypy will already catch this via the fact we typed `UserProfile` as `Optional[UserProfile]`.